From d1f044c9ed7ec0590f6d60b11a1cb0e3b13eafa8 Mon Sep 17 00:00:00 2001
From: Jean-Yves Avenard <jyavenard@mythtv.org>
Date: Sat, 28 Jun 2014 00:12:31 +1000
Subject: [PATCH 09/16] Fix PiP with VDPAU on mesa's drivers

This is due to a bug in mesa's vdpau implementation.
https://bugs.freedesktop.org/show_bug.cgi?id=80561

So as a work around, create a dummy bitmap surface and use this.
Limit the work around to non-nvidia drivers

There is still a remaining issue, only one PiP shows at a time. This is due to another bug in mesa's vdpau VdpVideoMixerRender, it clears the whole destination surface. So you only see the last PiP window.
---
 mythtv/libs/libmythui/mythrender_vdpau.cpp |   41 ++++++++++++++++++++++++++--
 1 file changed, 38 insertions(+), 3 deletions(-)

diff --git a/mythtv/libs/libmythui/mythrender_vdpau.cpp b/mythtv/libs/libmythui/mythrender_vdpau.cpp
index c021e3b..6094685 100644
--- a/mythtv/libs/libmythui/mythrender_vdpau.cpp
+++ b/mythtv/libs/libmythui/mythrender_vdpau.cpp
@@ -1383,7 +1383,6 @@ bool MythRenderVDPAU::DrawBitmap(uint id, uint target,
         vdest.y0 = (dst->y() < 0) ? 0 : dst->y();
         vdest.x1 = dst->x() + width;
         vdest.y1 = dst->y() + height;
-
     }
 
     if (src)
@@ -1404,12 +1403,48 @@ bool MythRenderVDPAU::DrawBitmap(uint id, uint target,
     }
 
     INIT_ST
+
+    bool createdBitmap = false;
+
+    if (!id && !gVDPAUNVIDIA)
+    {
+        // Work around MESA bug #80561
+        vdp_st = vdp_bitmap_surface_create(m_device, VDP_RGBA_FORMAT_B8G8R8A8,
+                                           1, 1, true,
+                                           &bitmap);
+        CHECK_ST
+
+        if (ok)
+        {
+            uint8_t bmp[] = { 255, 255, 255, 255 };
+            void *plane[1] = { bmp };
+            uint32_t pitch[1] = { 4 };
+            vdp_st =
+                vdp_bitmap_surface_put_bits_native(bitmap, plane, pitch, NULL);
+            CHECK_ST
+            if (!ok)
+            {
+                vdp_st = vdp_bitmap_surface_destroy(bitmap);
+                bitmap = VDP_INVALID_HANDLE;
+            }
+            else
+            {
+                createdBitmap = ok;
+            }
+        }
+    }
+
     vdp_st = vdp_output_surface_render_bitmap_surface(
-                surface,
-                dst ? &vdest : NULL, bitmap, src ? &vsrc  : NULL,
+                surface, dst ? &vdest : NULL, bitmap, src ? &vsrc  : NULL,
                 alpha >= 0 ? &color : NULL, &VDPBlends[blend],
                 VDP_OUTPUT_SURFACE_RENDER_ROTATE_0);
     CHECK_ST
+
+    if (createdBitmap)
+    {
+        vdp_st = vdp_bitmap_surface_destroy(bitmap);
+    }
+
     return ok;
 }
 
-- 
1.7.10.2

